### PR TITLE
refact gradient checkpoints

### PR DESCRIPTION
**Pull request by fsx950223**
Support training efficientdet-d7x(batch_size 2) by keras/train.py and Support training efficientdet-d6(batch_size 2) by main.py on GTX1080Ti(11G).
Thanks to @kartik4949 .

### CODE REVIEW

**Comment by kartik4949** [comment]
@fsx950223 thanks ,
And great job implementing gradient checkpointing with recomput_grad 

Lastly , as I discussed did you try it will other combinations like having a different  split ? 
Though current setup looks good to me.
:)

**Comment by fsx950223** [comment]
The solution splits gradient according to peak memory usage in the GPU.

**Comment by kartik4949** [comment]
Great , makes sense :)

**Review by kartik4949 — APPROVED** [review]
LGTM!

**Comment by dathudeptrai** [comment]
@fsx950223 will support this feature for TF2 ?

**Comment by fsx950223** [comment]
> @fsx950223 will support this feature for TF2 ?

Support training efficientdet-d7x(batch_size 2) by keras/train.py

File: efficientdet/backbone/efficientnet_model.py [hunk 1]
```diff
@@ -36,7 +36,7 @@
     'num_classes', 'width_coefficient', 'depth_coefficient', 'depth_divisor',
     'min_depth', 'survival_prob', 'relu_fn', 'batch_norm', 'use_se',
     'local_pooling', 'condconv_num_experts', 'clip_projection_output',
-    'blocks_args', 'fix_head_stem'
+    'blocks_args', 'fix_head_stem', 'gradient_checkpoints'
```
Comment (mingxingtan) [inline efficientdet/backbone/efficientnet_model.py:39]: rename "gradient_checkpoints" to "grad_checkpoint" for simplicity?

**Review by mingxingtan — APPROVED** [review]
Amazing! Thanks @fsx950223 !

**Comment by NikZak** [comment]
Amazing job!
@fsx950223, do you mind sharing what did you use to measure the peak memory usage nodes in the gpu? I would like to monitor that as well

File: efficientdet/keras/efficientdet_keras.py [hunk 2]
```diff
@@ -384,13 +385,14 @@ def __init__(self,
...
+    self.grad_checkpoint = grad_checkpoint
     if separable_conv:
       conv2d_layer = functools.partial(
           tf.keras.layers.SeparableConv2D,
           depth_multiplier=1,
           data_format=data_format,
-          pointwise_initializer=tf.initializers.VarianceScaling(),
-          depthwise_initializer=tf.initializers.VarianceScaling())
+          pointwise_initializer=tf.initializers.variance_scaling(),
+          depthwise_initializer=tf.initializers.variance_scaling())
```
Comment (dkurt) [inline efficientdet/keras/efficientdet_keras.py:395]: It makes the project incompatible with TensorFlow 2.1.0:
```
AttributeError: module 'tensorflow_core.python.keras.api._v2.keras.initializers' has no attribute 'variance_scaling'
```

**Comment by fsx950223** [comment]
> Amazing job!
> @fsx950223, do you mind sharing what tool did you use to measure the peak memory usage nodes in the gpu? I would like to monitor that as well

https://www.tensorflow.org/guide/profiler

File: efficientdet/keras/efficientdet_keras.py (hunk 2 above)
Comment (dkurt) [reply efficientdet/keras/efficientdet_keras.py:395]: /cc @fsx950223

Comment (fsx950223) [reply efficientdet/keras/efficientdet_keras.py:395]: Update to tf2.3

Comment (dkurt) [reply efficientdet/keras/efficientdet_keras.py:395]: @fsx950223, `requirements.txt` should be updated this way: https://github.com/google/automl/blob/3dd89c4b2443e8b8037018ff805edc1aa09fa6ee/efficientdet/requirements.txt#L7

Comment (fsx950223) [reply efficientdet/keras/efficientdet_keras.py:395]: > @fsx950223, `requirements.txt` should be updated this way:
> 
> https://github.com/google/automl/blob/3dd89c4b2443e8b8037018ff805edc1aa09fa6ee/efficientdet/requirements.txt#L7

Yeah, https://github.com/google/automl/pull/846

Comment (dkurt) [reply efficientdet/keras/efficientdet_keras.py:395]: Thank you!

**Pull request merged**